File: go/stack/valid_parentheses.go
package stack

func isValid(s string) bool {
	stack := []rune{}
	closeToOpen := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	for _, c := range s {
		if v, ok := closeToOpen[c]; ok {
			if len(stack) > 0 && stack[len(stack)-1] == v {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		} else {
			stack = append(stack, c)
		}
	}
	return len(stack) == 0
}

func IsValid(s string) bool {
	return isValid(s)
}

File: stack/valid_parentheses.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> stack;
        unordered_map<char, char> closeToOpen = {
            {')', '('},
            {']', '['},
            {'}', '{'}};

        for (char c : s)
        {
            if (closeToOpen.count(c))
            {
                if (!stack.empty() && stack.top() == closeToOpen[c])
                {
                    stack.pop();
                }
                else
                {
                    return false;
                }
            }
            else
            {
                stack.push(c);
            }
        }
        return stack.empty();
    }
};

int main()
{
    string input = "[(])";
    Solution *s = new Solution();
    bool output = s->isValid(input);
    cout << output << endl;
    return 0;
}